### ISSUE TITLE
Variable batch sizes in flax.nn.DenseGeneral

### ISSUE THREAD

**Issue by danielsuo** [google/flax#246]
Simple use case: I have some input data that the batch size does not divide evenly.

In `flax.nn.Dense`, the kernel shape is `(input.shape[-1], features)`, which means each batch item has the same kernel and we can pass arbitrarily-sized batches to the same model and get back

By contrast, `flax.nn.DenseGeneral` has a kernel shape `batch_shape + kernel_shape` which locks us into a batch size (see self-contained example below). We also train a factor of `batch_shape` more parameters as a result, where each item in a batch has its own copy of `kernel`.

This second point might not matter, but the behavior is different from `flax.nn.Dense`; I had previously thought `flax.nn.Dense` was implemented using `flax.nn.DenseGeneral`.

The first solution that comes to mind is we ignore the value in the batch dimensions when we initialize the kernel/bias parameters, much as `flax.nn.Dense` does. However, we would need to be careful if the user has no batch dimensions.

The above might be intended behavior, but just wanted to note my surprise / desire for something a little different.

```python
import flax
import jax
import numpy as np

model_def = flax.nn.DenseGeneral.partial(
    features=1,
    axis=(1, 2),
    batch_dims=0,
    kernel_init=flax.nn.initializers.kaiming_normal()
)

_, params = model_def.init_by_shape(jax.random.PRNGKey(0), [(2, 1, 57)])

print(params["bias"].shape, params["kernel"].shape)
# (2, 1) (2, 1, 57, 1)

X = np.random.rand(3, 1, 57)

model(X)
# ValueError: Existing shape (2, 1, 57, 1) differs from requested shape (3, 1, 57, 1)
```

**Comment by avital** [comment]
Hey @jheek or @levskaya -- could you take a look at this proposal? I'm not fully acquainted with all of the design decisions behind `flax.nn.DenseGeneral`.

**Comment by jheek** [comment]
This actually looks like a bug to me. Will investigate

**Comment by jheek** [comment]
Actually this is intended behavior. `batch_dims` refers to batched matmuls (see `jax.lax.dot_general`). So in this case you are computing the einsum `xyz, xyz->xyz` where you probably intented `xyz, yz->xyz`.
We could consider a different name for batched matmuls to avoid the terminology clash between linear algebra and ML.

**Comment by danielsuo** [comment]
Yeah, I did take a look at `jax.lax.dot_general` and the underlying `xla` docs, but they seemed to do what I expected (i.e., batched matmuls as you said).

What I didn't understand was the intended behavior of `DenseGeneral` since it appears to differ from `Dense`; are we saying we maintain `batch_dims` sets of parameters for the input? Separately, if we do want variable batch sizes, should we just implement another `flax.nn.Module`?

Fine with any of these, just wanted to make sure before we went ahead and wrote our own "VariableBatchDenseGeneralAwfulName"

**Comment by danielsuo** [comment]
Actually backing up a second; is `DenseGeneral` intended to be a `flax` wrapper around `dot_general`? That might explain my confusion.

**Comment by jheek** [comment]
Yes it is pretty much a wrapper around dot_general. The complexity arises due to kernel initialization. So to be clear
a batch axis in ML is an axis that only appears in the inputs
in a batched matmul the batch dim appears in both the input and the kernel.
So `axis` dims are contracted over, while `batch_dims` appear in both the input and the kernel, any remaining dimension that appear in the input will be multiplied with the same kernel (ML style batching)

**Comment by danielsuo** [comment]
Yep, makes sense! Once I realized `DenseGeneral` is essentially just a wrapper around `dot_general`, there was no more confusion. I'd still note my original surprise, but I'm not sure how big a deal that is vs. making a breaking change.

Anyway, we kept a version of my example implementation in `timecast` [here](https://github.com/MinRegret/timecast/blob/master/timecast/learners/_linear.py).

I'll close this issue and the associated PR.
